ClientRuntimes/CSharp/Microsoft.Rest.ClientRuntime/BatchDelegatingHandler.cs

```diff
@@ -0,0 +1,185 @@
+            }
+
+            if (!batchResponse.Content.IsMimeMultipartContent("mixed"))
+            {
+                throw new Exception("Batch response returned unexpected content type.");
+            }
+
+            // pull out the responses
+            MultipartMemoryStreamProvider responseContents = await batchResponse.Content.ReadAsMultipartAsync();
+            if (responseContents == null)
+            {
+                throw new Exception("Failed to convert response content into multipart components.");
+            }
+
+            if (responseContents.Contents.Count != this.requests.Count)
+            {
+                throw new Exception("Batch response returned " + responseContents.Contents.Count + " number of responses instead of the expected " + this.requests.Count + ".");
+            }
+
+            // copy the responses over into the array that is accessible by the tasks waiting for individual responses
+            this.responses = new HttpContent[responseContents.Contents.Count];
+            responseContents.Contents.CopyTo(this.responses, 0);
+
+            // signal tasks that they can proceed with using the response
+            this.batchCompleted.Set();
+        }
+
+        /// <summary>
+        /// Queue an HTTP request that will be eventually sent to the inner handler when you call IssueBatch.
+        /// DO NOT await nor access the result of the returned task until IssueBatch has been successfully awaited.
+        /// </summary>
+        /// <param name="request">HTTP request message to send to the server</param>
+        /// <param name="cancellationToken">Not used</param>
+        /// <returns>Task that when awaited will provide the HTTP response from the server</returns>
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            // Sanity check. There must not be any queued responses from the server.
+            if (this.responses != null)
+            {
+                throw new Exception("This batch handler instance has been reused without calling reset first.");
+            }
+
+            // what is my queue position?
+            int position = this.requests.Count;
+
+            // add the request to the queue
+            this.requests.Add(request);
+
+            // send back a task that will return the appropriate value
+            Task<HttpResponseMessage> task = Task<HttpResponseMessage>.Factory.StartNew(() => this.GetResponse(position));
+            return task;
+        }
+
+        /// <summary>
+        /// Wait for the batch call to finish and then return the requested response.
+        /// </summary>
+        /// <param name="index">Which response to return</param>
+        /// <returns>The HTTP response message inside the batch at the given index</returns>
+        private HttpResponseMessage GetResponse(int index)
+        {
+            this.batchCompleted.WaitOne();
```

Comment:
Having a timeout is a good idea. If you want the timeout on individual SendAsync calls, then the developer would have to factor in (1) the time between making individual SendAsync calls and the IssueBatch call; (2) time between the IssueBatch call and the response coming back & being parsed. Is that still useful, or would it make more sense to have the timeout on the IssueBatch call? In the latter case, I can add it as an optional parameter to the IssueBatch call.